server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java

```diff
@@ -111,19 +102,30 @@ private static LinearRing filterRing(LinearRing linearRing) {
         lons[0] = lons[count - 1] = linearRing.getLon(0);
         count = 0;
         for (int i = 1; i < numPoints - 1; i++) {
-            if (linearRing.getLon(i - 1) == linearRing.getLon(i)) {
-                if (linearRing.getLat(i - 1) == linearRing.getLat(i) || linearRing.getLon(i - 1) == linearRing.getLon(i + 1)) {
-                    // filter
-                    continue;
-                }
+            if (skipPoint(linearRing, i) == false) {
+                count++;
+                lats[count] = linearRing.getLat(i);
+                lons[count] = linearRing.getLon(i);
             }
-            count++;
-            lats[count] = linearRing.getLat(i);
-            lons[count] = linearRing.getLon(i);
         }
         return new LinearRing(lons, lats);
     }
 
+    private static boolean skipPoint(LinearRing linearRing, int i) {
+        if (linearRing.getLon(i - 1) == linearRing.getLon(i)) {
+            if (linearRing.getLat(i - 1) == linearRing.getLat(i)) {
+                // same point
+                return true;
+            }
+            if (linearRing.getLon(i - 1) == linearRing.getLon(i + 1)
+                && linearRing.getLat(i - 1) > linearRing.getLat(i) != linearRing.getLat(i + 1) > linearRing.getLat(i)) {
+                // coplanar
```

Comment:
I am a bit confused about terminology and the logic here. We have only 2 coordinates, so theoretically all our points are coplanar. Do you mean collinear here? Based on this logic we only remove points if we go down and then up within the same longitude. What's the reason for not removing colinear points in between? Basically, in the example below: `POLYGON ((0 0, 0 10, 10 10, 10 5, 10 -10, 10 0, 0 0))` the current logic removes point B, but keep point A intact. 

![image](https://user-images.githubusercontent.com/655851/144146805-2b75fa11-e4bf-4bdd-af6a-2d53223ad875.png)

Reply:
doh! Yes, I meant collinear points

Reply:
I had a look and the current logic that what I expect that is to remove point A:

original:

<img width="318" alt="image" src="https://user-images.githubusercontent.com/29038686/144223858-14b3d99f-17ef-460d-933b-9b72db4bd440.png">

processed:

<img width="311" alt="image" src="https://user-images.githubusercontent.com/29038686/144223961-ac4de766-0f4e-460b-ba7c-08f970db39e0.png">